### DISCUSSION TITLE
MacOS: Compliation of Jinx failed. Issue related to Homebrew?

### DISCUSSION THREAD

**Discussion by alecStewart1** [minad/jinx#160] [Q&A] [answered]
Hello!

For awhile, Jinx was working fine, and until the most recent version. I get the follow error:

```
gcc -I. -O2 -Wall -Wextra -fPIC -shared -o jinx-mod.dylib jinx-mod.c -I/usr/include/enchant-2 -lenchant-2
jinx-mod.c:19:10: fatal error: 'enchant.h' file not found
#include <enchant.h>
         ^~~~~~~~~~~
1 error generated.
Jinx: Compilation of jinx-mod.dylib failed
```

I've instead enchant via Homebrew:

```
$ homebrew info enchant
https://abiword.github.io/enchant/
/usr/local/Cellar/enchant/2.6.8 (29 files, 297.8KB) *
  Poured from bottle using the formulae.brew.sh API on 2024-03-25 at 09:27:02
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/e/enchant.rb
License: LGPL-2.1-or-later
==> Dependencies
Build: pkg-config ✔
Required: aspell ✔, glib ✔
==> Analytics
install: 3,589 (30 days), 11,350 (90 days), 34,394 (365 days)
install-on-request: 1,909 (30 days), 6,053 (90 days), 18,050 (365 days)
build-error: 0 (30 days)
```

And `enchant.h` seems to be in the following places:
```
$ fd -t f 'enchant.h' /
/usr/local/Cellar/enchant/2.6.8/share/doc/enchant/enchant.html
/usr/local/Cellar/enchant/2.6.8/include/enchant-2/enchant.h
/System/Volumes/Data/usr/local/Cellar/enchant/2.6.8/share/doc/enchant/enchant.html
/System/Volumes/Data/usr/local/Cellar/enchant/2.6.8/include/enchant-2/enchant.h
```

So `/usr/local/Cellar/enchant/2.6.8/include/enchant-2/enchant.h` and `/System/Volumes/Data/usr/local/Cellar/enchant/2.6.8/include/enchant-2/enchant.h`. 

I don't know what changed in the newest version of Jinx, but I've just now ran into this issue.

**Comment by alecStewart1** [comment]
Oh, this could be an Emacs 30.0.50 thing, too.

**Comment by J-Loudet** [comment]
If that can help anyone, I manually compiled the shared library doing the following steps:
- in `jinx-mode.c` I replaced the line `#include <enchant.h>` with `#include "/opt/homebrew/include/enchant-2/enchant.h"`,
- I changed the compile command to: `gcc -I. -O2 -Wall -Wextra -fPIC -shared -o jinx-mod.dylib jinx-mod.c -L/opt/homebrew/Cellar/enchant/2.6.8/lib/ -lenchant-2`

These are obviously quick and dirty hacks…

**Reply by alecStewart1** [reply]
Maybe you set something with homebrew differently than me, but I believe the header file should be at 

`/usr/local/Cellar/enchant/2.6.8/include/enchant-2/enchant.h`

Is there not some variable in Emacs or some environment variable I'm forgetting about to add to the path of possible header files to include?

**Reply by alecStewart1** [reply]
Answered my own question. There is `C_INCLUDE_PATH`.

https://gcc.gnu.org/onlinedocs/cpp/Environment-Variables.html

**Reply by alecStewart1** [reply]
Alright, I don't know if this is helpful, but if gcc does look at `C_INCLUDE_PATH` before compiling things, you /could/ do this:

You can get the current version of `enchant` that Homebrew has installed:

```sh
brew info --json enchant | jq -c '.[].versions.stable'
```

We know that the header file will be at `/usr/local/Cellar/enchant/x.x.x/include/enchant-2/enchant.h`, so we could set the `C_INCLUDE_PATH` **in Emacs** to try and resolve this for now. So:

```emacs-lisp
(use-package jinx
  :preface
  (defvar homebrew-enchant-version
    (shell-command-to-string "brew info --json enchant | jq -c '.[].versions.stable'")
    "Capture the current version of enchant installed by Homebrew.")

  (defvar homebrew-enchant-include-path
    (concat "/usr/local/Cellar/enchant/" homebrew-enchant-version "/include/enchant-2/")
    "The path to the header files for enchant.")
  :init
  ;; I'm assuming paths in this environment variable are separated by a colon 
  (setenv "C_INCLUDE_PATH" (concat (getenv "C_INCLUDE") ":" homebrew-enchant-include-path))
  :config 
  ;; the rest of your config here
  )
```

Maybe that works, but I'll have to try it out.

**Reply by alecStewart1** [reply]
Oh, I'm an idiot. It's because the command to compile the jinx library is 

```
gcc -I. -O2 -Wall -Wextra -fPIC -shared -o jinx-mod.dylib jinx-mod.c -I/usr/include/enchant-2 -lenchant-2
```

So you'd have to change the `-I/usr/include/enchant-2` flag somehow.

**Reply by alecStewart1** [reply]
Also (we're getting into some weirdness here) if I run

```sh
pkg-config --list-all | rg 'enchant'
```

I get

```
enchant                                     libenchant - A spell checking library
enchant-2                                   libenchant - A spell checking library
```

So...I don't really know what Emacs is doing. I guess the `-I/usr/include/enchant-2` flag/option is what's throwing things off.

**Reply by J-Loudet** [reply]
> Maybe you set something with homebrew differently than me, but I believe the header file should be at
> 
> `/usr/local/Cellar/enchant/2.6.8/include/enchant-2/enchant.h`

I put that path as it's shorter (😅) and, on my machine©, it is a symlink to `/opt/homebrew/Cellar/enchant/2.6.8/include/enchant-2`. I do not have a `/usr/local/Cellar` folder.

I haven't been a macOS user for long so I don't know the intricacies of homebrew.

**Reply by minad** [reply]
Generally, the problem seems to be that pkgconf/pkg-config is not executed correctly on your system or doesn't find Enchant. Then as a last resort, Jinx falls back to the "default path" of "/usr/include/enchant-2", which then fails.

**Reply by paskn** [reply]
FYI, a similar problem seems to have happened with PDF-tools after some change in how brew installs packages  https://github.com/vedang/pdf-tools/issues/269#issuecomment-2039758759

I got to this discussion after I restarted Emacs and jinx greeted me with 'enchant.h' file not found 😅

**Reply by alecStewart1** [reply]
Ah, I think I found one thing. Check `/usr/local/lib/pkgconfig` on your system and if `PKG_CONFIG_PATH` has that path.

**EDIT 1:** Nevermind. I think I need to try reinstalling Emacs first. For me `/usr/local/lib/pkgconfig` has a lot of symlinks to the pkg-config files that homebrew installs along with a lot of packages. I'll see if after reinstalling Emacs if this fixes things.

**EDIT 2:** Nope.

**Reply by paskn** [reply]
This worked for me. Since I use `straight`, I had to do one more step by linking the compiled `dylib` to `/straight/build/jinx/`

> If that can help anyone, I manually compiled the shared library doing the following steps:
> 
>     * in `jinx-mode.c` I replaced the line `#include <enchant.h>` with `#include "/opt/homebrew/include/enchant-2/enchant.h"`,
> 
>     * I changed the compile command to: `gcc -I. -O2 -Wall -Wextra -fPIC -shared -o jinx-mod.dylib jinx-mod.c -L/opt/homebrew/Cellar/enchant/2.6.8/lib/ -lenchant-2`
> 
> 
> These are obviously quick and dirty hacks…

**Comment by J-Loudet** [comment, marked as answer]
Inspired by this [commit](https://github.com/vedang/pdf-tools/commit/a1048bceb2bd3f635437b0f2bfac27cae8c2dabc), adding the following lines to my `use-package` declaration seems to be enough to solve the compilation issue without having to manually modify files:

```emacs-lisp
  :init
  (setenv "PKG_CONFIG_PATH" (concat "/opt/homebrew/opt/glib/lib/pkgconfig/:" (getenv "PKG_CONFIG_PATH")))
```

**Reply by ChetanKoneru** [reply]
Thank you for this. Right now this is the easiest workaround.

**Reply by alecStewart1** [reply]
This works for me as well.

**Reply by alkurbatov** [reply]
Worked for me too! But only with `pkg-config`.

**Reply by minad** [reply]
Is there something which should be updated in the Jinx README?

**Reply by J-Loudet** [reply]
@minad I commented out my `:init` section that updated the `PKG_CONFIG_PATH` environment variable before updating Jinx and it was successfully compiled. I am thus assuming that the issue got fixed upstream and there is no need to update the README. 😉

**Comment by cacology** [comment]
I ran across the same error  because of conflicting `pkg-config` installations from MacPorts and homebrew. I'm recording it here just to help someone else if they have the same issue:

Calling `which pkg-config` gave me `/opt/local/bin/pkg-config` (MacPorts) rather than `/opt/homebrew/bin/pkg-config` (homebrew). It was easy to correct my setup by changing my .zprofile from:

```
export PATH="/opt/local/bin:/opt/local/sbin:$PATH"
```

To
```
export PATH="$PATH:/opt/local/bin:/opt/local/sbin"
```

This will probably prevent some similar issues later on too. (FYI I only want MacPorts for the libraries for cyrus-sasl xoauth2, so I can use my work's Exchange server with mbsync. I haven't been able to figure out how to do this with homebrew yet.)
